mixer/pkg/protobuf/yaml/encoder.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package yaml
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+)
+
+type (
+	bufferWriter struct {
+		resolver    Resolver
+		buffer      *proto.Buffer
+		walker      Walker
+		skipUnknown bool
+	}
+	// Encoder transforms yaml that represents protobuf data into []byte
+	Encoder struct {
+		walker   Walker
+		resolver Resolver
+	}
+)
+
+// NewEncoder creates a Encoder
+func NewEncoder(fds *descriptor.FileDescriptorSet) *Encoder {
+	resolver := NewResolver(fds)
+	return &Encoder{resolver: resolver, walker: NewWalker(resolver)}
+}
+
+// EncodeBytes creates []byte from a yaml representation of a proto.
+func (c *Encoder) EncodeBytes(data map[interface{}]interface{}, msgName string, skipUnknown bool) ([]byte, error) {
+	buf := GetBuffer()
+	msgWriter := newBufferWriter(c.resolver, buf, c.walker, skipUnknown)
+	err := c.walker.Walk(data, msgName, skipUnknown, msgWriter)
+	bytes := buf.Bytes()
+	PutBuffer(buf)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+func newBufferWriter(resolver Resolver, buffer *proto.Buffer, walker Walker, skipUnknown bool) *bufferWriter {
+	return &bufferWriter{resolver: resolver, buffer: buffer, walker: walker, skipUnknown: skipUnknown}
+}
+
+func (w *bufferWriter) Visit(name string, data interface{}, field *descriptor.FieldDescriptorProto) (Visitor, bool, error) {
+	switch field.GetType() {
+	case descriptor.FieldDescriptorProto_TYPE_STRING:
+		v, ok := data.(string)
+		if !ok {
+			return nil, false, badTypeError(name, "string", data)
+		}
+
+		_ = w.buffer.EncodeVarint(encodeIndexAndType(int(field.GetNumber()), proto.WireBytes))
+		_ = w.buffer.EncodeStringBytes(v)
+	case descriptor.FieldDescriptorProto_TYPE_DOUBLE:
+		v, ok := data.(float64)
+		if !ok {
+			return nil, false, badTypeError(name, "float64", data)
+		}
+		_ = w.buffer.EncodeVarint(encodeIndexAndType(int(field.GetNumber()), proto.WireFixed64))
```

Comment:
In general, it would be useful if you can add a comment about why the errors from encode are ignored.

Reply:
D